cpp/src/arrow/scalar.cc

```diff
@@ -252,6 +270,100 @@ Result<std::shared_ptr<Scalar>> Scalar::Parse(const std::shared_ptr<DataType>& t
   return ScalarParseImpl{type, s}.Finish();
 }
 
+struct ScalarFromArraySlotImpl {
+  template <typename T>
+  using ScalarType = typename TypeTraits<T>::ScalarType;
+
+  Status Visit(const NullArray& a) {
+    out_ = std::make_shared<NullScalar>();
+    return Status::OK();
+  }
+
+  Status Visit(const BooleanArray& a) { return Finish(a.Value(index_)); }
+
+  template <typename T>
+  Status Visit(const NumericArray<T>& a) {
+    return Finish(a.Value(index_));
+  }
+
+  template <typename T>
+  Status Visit(const BaseBinaryArray<T>& a) {
+    return Finish(a.GetString(index_));
```

Comment:
Use `GetView` since you copy it right after in a Buffer, that'll remove one memcpy.

Reply:
will do

Reply:
Buffer::FromString wraps the allocation of the std::string without copying